### PR TITLE
Add C implementation for `@stdlib/math/base/special/kernel-tan`

### LINKED ISSUE #646: [RFC]: Add C implementation for `@stdlib/math/base/special/kernel-tan`
### Description

This RFC proposes adding a C implementation for [`@stdlib/math/base/special/kernel-tan`](https://github.com/stdlib-js/stdlib/blob/b43a75923e47775446d3b6d7219ee42e07f7ce63/lib/node_modules/%40stdlib/math/base/special/kernel-tan).

```c
double stdlib_base_kernel_tan( const double x, const double y, const int32_t k );
```

### Related Issues

None.

### Questions

No.

### Other

No.

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by Pranavchiku**
Resolves #646  .

## Checklist

- [x] update readme.md
- [x] include.gypi
- [x] binding.gyp
- [x] include/stdlib/math/base/special/
- [x] src
- [x] manifest.json
- [x] lib
- [x] examples
- [x] benchmark
- [x] test
* * *

@kgryte

### CODE REVIEW

**Comment by Pranavchiku** [comment]
Some of the tests are failing, I will have a look at those asap.

**Comment by Pranavchiku** [comment]
198/3008 tests are failing and are like 
```
    operator: equal
    expected: |-
      -5.281879551675655
    actual: |-
      -5.2818795516756545
```
I don't know what's the problem here, Implementations look fine, addon.c looks good to me. @kgryte if you can have a look that would be of great help.

**Comment by kgryte** [comment]
It could be floating-point rounding error. Will take a look.

File: lib/node_modules/@stdlib/math/base/special/kernel-tan/src/main.c [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	double v;
+	double w;
+	double s;
+	double xc = x;
+	double yc = y;
+
+	stdlib_base_float64_get_high_word( x, &hx );
+
+	// High word of |x|:
+	ix =  ( int32_t ) ( ( int32_t ) hx & HIGH_WORD_ABS_MASK );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/kernel-tan/src/main.c]: The `(int32_t)` casts are not necessary here.

Comment (Pranavchiku) [reply lib/node_modules/@stdlib/math/base/special/kernel-tan/src/main.c]: Removed those, still tests are failing

**Comment by kgryte** [comment]
@Pranavchiku Issue seems to be two bugs in the JS implementation. Try fixing those in this PR and running the tests again.

**Comment by Pranavchiku** [comment]
All tests and benchmarks passed, this PR can be reviewed.

File: lib/node_modules/@stdlib/math/base/special/kernel-tan/README.md [hunk 2]
```diff
@@ -120,6 +120,105 @@ for ( i = 0; i < x.length; i++ ) {
...
+
+### Usage
+
+```c
+#include "stdlib/math/base/special/kernel_tan.h"
+```
+
+#### stdlib_base_kernel_tan( x, y, k)
+
+Computes the [tangent][tangent] of a `number` on `[-π/4, π/4]`. For increased accuracy, the number for which the [tangent][tangent] should be evaluated can be supplied as a [double-double number][double-double-arithmetic] (i.e., a non-evaluated sum of two [double-precision floating-point numbers][ieee754] `x` and `y`).
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/kernel-tan/README.md:151]: Move to note.

File: lib/node_modules/@stdlib/math/base/special/kernel-tan/README.md [hunk 3]
```diff
@@ -120,6 +120,105 @@ for ( i = 0; i < x.length; i++ ) {
...
+
+out = stdlib_base_kernel_tan( 3.141592653589793/6.0, 0.0, 1 );
+// returns ~0.577
+```
+
+The function accepts the following arguments:
+
+-   **x**: `[in] double`  input value (in radians, assumed to be bounded by `~pi/4` in magnitude).
+-   **y**: `[in] double`  tail of `x`.
+-   **k**: `[in] int32_t` if `k = 1`, the function returns `tan(x+y)`. To return the negative inverse `-1/tan(x+y)`, set `k = -1`.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/kernel-tan/README.md:165]:
```suggestion
-   **k**: `[in] int32_t` indicates whether `tan(x+y)` (if `k = 1`) or `-1/tan(x+y)` (if `k = -1`) is returned.
```

File: lib/node_modules/@stdlib/math/base/special/kernel-tan/README.md [hunk 4]
```diff
@@ -120,6 +120,105 @@ for ( i = 0; i < x.length; i++ ) {
...
+
+<!-- C API usage examples. -->
+
+<section class="examples">
+
+### Examples
+
+```c
+#include "stdlib/math/base/special/kernel_tan.h"
+#include <stdlib.h>
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/kernel-tan/README.md:191]: Needs updating.

File: lib/node_modules/@stdlib/math/base/special/kernel-tan/README.md [hunk 5]
```diff
@@ -120,6 +120,105 @@ for ( i = 0; i < x.length; i++ ) {
...
+#include <stdint.h>
+#include <stdio.h>
+#include <math.h>
+#include <inttypes.h>
+
+int main() {
+    double out;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/kernel-tan/README.md:198]: Need to reorder.

File: lib/node_modules/@stdlib/math/base/special/kernel-tan/README.md [hunk 6]
```diff
@@ -120,6 +120,105 @@ for ( i = 0; i < x.length; i++ ) {
...
+    double x[] = { -0.7853981633974483, -0.6108652381980153, -0.4363323129985824, -0.26179938779914946, -0.08726646259971649, 0.08726646259971649,
+    0.26179938779914935, 0.43633231299858233, 0.6108652381980153, 0.7853981633974483 };
+    double y;
+    int32_t k;
+    int i;
+
+    for ( i = 0; i < 10; i++ ) {
+        y = 0.0;
+        k = 1;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/kernel-tan/README.md:207]: Can simplify.

File: lib/node_modules/@stdlib/math/base/special/kernel-tan/benchmark/c/benchmark.c [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+	return (double)r / ( (double)RAND_MAX + 1.0 );
+}
+
+/**
+* Runs a benchmark.
+*
+* @return elapsed time in seconds
+*/
+double benchmark() {
+	int32_t k;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/kernel-tan/benchmark/c/benchmark.c:96]: Reorder.

File: lib/node_modules/@stdlib/math/base/special/kernel-tan/benchmark/c/benchmark.c [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+	double y;
+	double z;
+	double t;
+	int i;
+
+	t = tic();
+	for ( i = 0; i < ITERATIONS; i++ ) {
+		x = ( ( rand_double() * 2.0 ) - 1.0 ) * 0.7853981633974483;
+		y = 0.0;
+		k = 1;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/kernel-tan/benchmark/c/benchmark.c:108]: Can be simplified.

File: lib/node_modules/@stdlib/math/base/special/kernel-tan/examples/c/example.c [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+
+#include "stdlib/math/base/special/kernel_tan.h"
+#include <stdlib.h>
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/kernel-tan/examples/c/example.c:21]: Need to clean-up.

File: lib/node_modules/@stdlib/math/base/special/kernel-tan/examples/c/example.c [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+#include <stdint.h>
+#include <stdio.h>
+#include <math.h>
+#include <inttypes.h>
+
+int main() {
+    double out;
+    double x[] = { -0.7853981633974483, -0.6108652381980153, -0.4363323129985824, -0.26179938779914946, -0.08726646259971649, 0.08726646259971649,
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/kernel-tan/examples/c/example.c:29]: Need to reorder.

File: lib/node_modules/@stdlib/math/base/special/kernel-tan/examples/c/example.c [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    0.26179938779914935, 0.43633231299858233, 0.6108652381980153, 0.7853981633974483 };
+    double y;
+    int32_t k;
+    int i;
+
+    for ( i = 0; i < 10; i++ ) {
+        y = 0.0;
+        k = 1;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/kernel-tan/examples/c/example.c:37]: Can simplify.

File: lib/node_modules/@stdlib/math/base/special/kernel-tan/src/addon.c [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include "stdlib/math/base/special/kernel_tan.h"
+#include <node_api.h>
+#include <stdint.h>
+#include <stdbool.h>
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/kernel-tan/src/addon.c:22]: Can remove.

File: lib/node_modules/@stdlib/math/base/special/kernel-tan/src/main.c [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include "stdlib/math/base/special/kernel_tan.h"
+#include "stdlib/number/float64/base/get_high_word.h"
+#include "stdlib/number/float64/base/set_low_word.h"
+#include "polyval_t_even.c"
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/kernel-tan/src/main.c:22]: Refactor to inline polynomial functions.

File: lib/node_modules/@stdlib/math/base/special/kernel-tan/src/main.c [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+#include "polyval_t_odd.c"
+
+static const double PI04 = 7.85398163397448278999e-01;
+static const double PI04LO = 3.06161699786838301793e-17;
+static const double T0 = 3.33333333333334091986e-01; // 3FD55555, 55555563
+
+// Absolute value mask: 2147483647 => 0x7fffffff => 01111111111111111111111111111111
+static const int32_t HIGH_WORD_ABS_MASK = 0x7fffffff;
+
+
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/kernel-tan/src/main.c:32]:
```suggestion
```

File: lib/node_modules/@stdlib/math/base/special/kernel-tan/src/main.c [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+/**
+* Computes the tangent of a number on [-π/4, π/4].
+*
+* @param x    input value (in radians, assumed to be bounded by `~pi/4` in magnitude)
+* @param y    tail of `x`
+* @param k    if `k = 1`, the function returns `tan(x+y)`. To return the negative inverse `-1/tan(x+y)`, set `k = -1`.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/kernel-tan/src/main.c:38]: Update to match JS

File: lib/node_modules/@stdlib/math/base/special/kernel-tan/src/main.c [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+* @return	  tangent value
+*
+* @example
+* double x = 1.0;
+* double y = 2.0;
+* int32_t k = 1;
+*
+* double out = stdlib_base_kernel_tan( x, y, k );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/kernel-tan/src/main.c:46]: Can simplify.

File: lib/node_modules/@stdlib/math/base/special/kernel-tan/src/main.c [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+*/
+double stdlib_base_kernel_tan( const double x, const double y, const int32_t k ) {
+	int32_t ix;
+	uint32_t hx;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/kernel-tan/src/main.c:50]: Reorder.

File: lib/node_modules/@stdlib/math/base/special/kernel-tan/src/main.c [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+		return ( 1.0 - ( (hx >> 30) & 2 ) ) * ( v - (2.0 * (xc - ((w * w / (w + v)) - r)) ));
+	}
+	if ( k == 1 ) {
+		return w;
+	}
+	// Compute -1/(x+r) accurately...
+	z = w;
+	stdlib_base_float64_set_low_word( 0, &z );
+	v = r - (z - xc); // z + v = r + xc
+	double a = -1.0 / w;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/kernel-tan/src/main.c:97]: Fix declaration.

File: lib/node_modules/@stdlib/math/base/special/kernel-tan/test/test.native.js [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+tape( 'the function can be used to compute the tangent for input values outside of `[-pi/4, pi/4]` after argument reduction via `rempio2` (positive)', opts, function test( t ) {
+	var values;
+	var out;
+	var x;
+	var y;
+	var n;
+	var i;
+
+	values = linspace( 40.0*PI/4.0, 200*PI/4.0, 1000 );
+	y = new Array( 2 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/kernel-tan/test/test.native.js:95]: Update to avoid constructor.

**Review by kgryte — APPROVED** [review]
As none of the requested changes are blocking, will merge and touch up on my end. Thanks, @Pranavchiku!

**Pull request merged**